tests/test_subtractTask.py

```diff
@@ -0,0 +1,459 @@
+                                    xSize=xSize, ySize=ySize, x0=x0, y0=y0)
+        params = [2.2, 2.1, 2.0, 1.2, 1.1, 1.0]
+
+        bbox2D = lsst.geom.Box2D(lsst.geom.Point2D(x0, y0), lsst.geom.Extent2D(xSize, ySize))
+        background_model = Chebyshev1Function2D(params, bbox2D)
+        science, sources = makeTestImage(psfSize=2.0, noiseLevel=noiseLevel, noiseSeed=6,
+                                         background=background_model,
+                                         xSize=xSize, ySize=ySize, x0=x0, y0=y0)
+        config = subtractImages.AlardLuptonSubtractTask.ConfigClass()
+        config.doSubtractBackground = True
+
+        config.makeKernel.kernel.name = "AL"
+        config.makeKernel.kernel.active.fitForBackground = True
+        config.makeKernel.kernel.active.spatialKernelOrder = 1
+        config.makeKernel.kernel.active.spatialBgOrder = 2
+        for mode in ["convolveTemplate", "convolveScience"]:
+            config.mode = mode
+            task = subtractImages.AlardLuptonSubtractTask(config=config)
+            output = task.run(template, science, sources)
+
+            # We should be fitting the same number of parameters as were in the input model
+            self.assertEqual(output.backgroundModel.getNParameters(), background_model.getNParameters())
+
+            # The parameters of the background fit should be close to the input model
+            self.assertFloatsAlmostEqual(np.array(output.backgroundModel.getParameters()),
+                                         np.array(params), rtol=0.3)
+
+            # stddev of difference image should be close to expected value.
+            # This will fail if we have mis-subtracted the background.
+            self.assertFloatsAlmostEqual(np.nanstd(output.difference.image.array),
+                                         np.sqrt(2)*noiseLevel, rtol=0.1)
+
+    def test_scale_variance(self):
+        """Check variance scaling of the image difference.
+        """
+        noiseLevel = 1.
+        scaleFactor = 2.345
+        science, sources = makeTestImage(psfSize=3.0, noiseLevel=noiseLevel, noiseSeed=6)
+        template, _ = makeTestImage(psfSize=2.0, noiseLevel=noiseLevel, noiseSeed=7, templateBorderSize=20)
+        config = subtractImages.AlardLuptonSubtractTask.ConfigClass()
+        scienceVarianceOrig = science.variance.array[...]
+        templateVarianceOrig = template.variance.array[...]
+        for scaleFactor in [1., scaleFactor]:
+            science.variance.array[...] = scienceVarianceOrig/scaleFactor
+            template.variance.array[...] = templateVarianceOrig/scaleFactor
+            scienceVarMean = np.mean(science.variance.array)
+            templateVarMean = np.mean(template.variance.array)
+            config.doSubtractBackground = False
+            for doDecorrelation in [True, False]:
+                for doScaleVariance in [True, False]:
+                    config.doDecorrelation = doDecorrelation
```

Comment:
This inner block should be a separate `_check_???` method called with all of the loop variables as args: otherwise it can be difficult to determine which branch of the loop actually failed.